supervisor/shared/safe_mode.c

```diff
@@ -103,77 +113,80 @@ void print_safe_mode_message(safe_mode_t reason) {
         return;
     }
     serial_write("\n");
-    // Output a user safe mode string if it's set.
-    #ifdef BOARD_USER_SAFE_MODE
-    if (reason == USER_SAFE_MODE) {
-        serial_write_compressed(translate("You requested starting safe mode by "));
-        serial_write(BOARD_USER_SAFE_MODE_ACTION);
-        serial_write_compressed(translate("\nTo exit, please reset the board without "));
-        serial_write(BOARD_USER_SAFE_MODE_ACTION);
-        serial_write("\n");
-    } else
-    #endif
-        switch (reason) {
-            case MANUAL_SAFE_MODE:
-                serial_write_compressed(translate("CircuitPython is in safe mode because you pressed the reset button during boot. Press again to exit safe mode.\n"));
-                return;
-            case PROGRAMMATIC_SAFE_MODE:
-                serial_write_compressed(translate("The `microcontroller` module was used to boot into safe mode. Press reset to exit safe mode.\n"));
-                return;
-            default:
-                break;
-        }
 
-        serial_write_compressed(translate("You are in safe mode: something unanticipated happened.\n"));
-        switch (reason) {
-            case BROWNOUT:
-                serial_write_compressed(translate("The microcontroller's power dipped. Make sure your power supply provides\nenough power for the whole circuit and press reset (after ejecting CIRCUITPY).\n"));
-                return;
-            case HEAP_OVERWRITTEN:
-                serial_write_compressed(translate("The CircuitPython heap was corrupted because the stack was too small.\nPlease increase the stack size if you know how, or if not:"));
-                serial_write_compressed(FILE_AN_ISSUE);
-                return;
-            case NO_HEAP:
-                serial_write_compressed(translate("CircuitPython was unable to allocate the heap.\n"));
-                serial_write_compressed(FILE_AN_ISSUE);
-                return;
-            default:
-                break;
-        }
+    switch (reason) {
+        case USER_SAFE_MODE:
+            #ifdef BOARD_USER_SAFE_MODE_ACTION
+                // Output a user safe mode string if it's set.
+                serial_write_compressed(translate("You requested starting safe mode by "));
+                serial_write_compressed(translate(BOARD_USER_SAFE_MODE_ACTION));
+                serial_write_compressed(translate("To exit, please reset the board without "));
+                serial_write_compressed(translate(BOARD_USER_SAFE_MODE_ACTION));
+            #else
+                // fallthrough
+                serial_write_compressed(translate("CircuitPython is in safe mode with status being USER_SAFE_MODE but no specific reason was provided.\n"));
```

Comment:
I'd prefer to see no extra text written here.  The reason is that string storage is precious, and I am not confident that the optimizer will eliminate the string by determining that `USER_SAFE_MODE` is an impossible value for `reason`.  I intended that when USER_SAFE_MODE was entered but there was no BOARD_USER_SAFE_MODE_ACTION, the MANUAL_SAFE_MODE message would be displayed, but I got this wrong in the code I shared on discord.  A second alternative would be to re-use the "Unknown reason" string from far down below.

It looks like I got what I wrote on discord the other day incorrect (the code needs to `return`, not `break` to end the processing of different safe-mode reasons), and I didn't describe in detail what I thought should happen when USER_SAFE_MODE was entered but there was no BOARD_USER_SAFE_MODE_ACTION.  I think the code should have said
```
    switch (reason) {
        case USER_SAFE_MODE:
            #ifdef BOARD_USER_SAFE_MODE_ACTION
            serial_write_compressed(translate("You requested starting safe mode by "));
            serial_write(BOARD_USER_SAFE_MODE_ACTION);
            serial_write_compressed(translate("\nTo exit, please reset the board without "));
            serial_write(BOARD_USER_SAFE_MODE_ACTION);
            serial_write("\n");
            return;   // <-- return instead of break
            #else
            // fallthrough
            #endif
        case MANUAL_SAFE_MODE:
            serial_write_compressed(translate("CircuitPython is in safe mode because you pressed the reset button during boot. Press again to exit safe mode.\n"));
```
The comment "fallthrough" is a hint for the reader of the code that (when the return statement isn't encountered) execution from the `case USER_SAFE_MODE` will proceeed through the lines for `MANUAL_SAFE_MODE`.  this is often called "falling through" a case. If the version of the function we end up taking has the `return` in every case before `case MANUAL_SAFE_MODE`, the fallthrough comment should be removed.  Not familiar with what fallthrough is in C?  This looks like an introduction that might help get you up to speed: https://en.wikipedia.org/wiki/Switch_statement#Fallthrough -- note that in CircuitPython I don't think we're careful to include these comments in general.

Reply:
@jepler my bad, I got the `fallthrough` definition wrong. Thanks for pointing that out.   

Considering what you said about string storage, I implemented the following change to get to the `Unknown Reason` output.

```suggestion
                break;
```
Following is the serial console output: 

```
You are in safe mode: something unanticipated happened. 
CircuitPython core code crashed hard. Whoops!
Unknown reason.
Please file an issue with the contents of your CIRCUITPY drive at
https://github.com/adafruit/circuitpython/issues
```

Reply:
Sounds good, remove the fallthrough comment too and then accept your own suggestion and this should be good to go.